### PR TITLE
Add MPG reward

### PR DESCRIPTION
**Pull request by eugenevinitsky**
Adds an MPG and MP-joule reward

### CODE REVIEW

File: flow/core/rewards.py [hunk 1]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
         power += M * speed * accel + M * g * Cr * speed + 0.5 * rho * A * Ca * speed ** 3
 
     return -gain * power
+
+
+def veh_energy_consumption(env, veh_id, gain=.001):
+    """Calculate power consumption of a vehicle.
+    Assumes vehicle is an average sized vehicle.
+    The power calculated here is the lower bound of the actual power consumed
+    by a vehicle.
```
Comment (Yasharzf) [inline flow/core/rewards.py:339]: add description of parameters and returns

File: flow/core/rewards.py [hunk 2]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+    Cr = 0.005  # rolling resistance coefficient
+    Ca = 0.3  # aerodynamic drag coefficient
+    rho = 1.225  # air density (kg/m^3)
+    A = 2.6  # vehicle cross sectional area (m^2)
+    speed = env.k.vehicle.get_speed(veh_id)
+    prev_speed = env.k.vehicle.get_previous_speed(veh_id)
+
+    accel = abs(speed - prev_speed) / env.sim_step
+
+    power += M * speed * accel + M * g * Cr * speed + 0.5 * rho * A * Ca * speed ** 3
```
Comment (Yasharzf) [inline flow/core/rewards.py:354]: do you have a reference especially for some of the coefficients (e.g., Cr)? if yes, you can add it to the docstring

File: flow/core/rewards.py [hunk 3]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+
+    return -gain * power
+
+
+def miles_per_megajoule(env, veh_id=None, gain=.001):
+    """Calculate miles per mega-joule of either a particular vehicle or the total average of all the vehilces.
+    Assumes vehicle is an average sized vehicle.
+    The power calculated here is the lower bound of the actual power consumed
+    by a vehicle.
```
Comment (Yasharzf) [inline flow/core/rewards.py:363]: params, returns?

File: flow/core/rewards.py [hunk 4]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+    counter = 0
+    if not isinstance(veh_id, list):
+        speed = env.k.vehicle.get_speed(veh_id)
+        # convert to be positive since the function called is a penalty
+        power = -veh_energy_consumption(env, veh_id, gain=1.0)
+        if power > 0 and speed >= 0.0:
+            # meters / joule is (v * \delta t) / (power * \delta t)
+            mpj = speed / power
+    else:
+        for veh_id in env.k.vehicle.get_ids():
```
Comment (Yasharzf) [inline flow/core/rewards.py:375]: shouldn't you iterate over vehicle.get_ids() if veh_id is None? here veh_id can be a list, shouldn't you iterate over veh_id list?

File: flow/core/rewards.py [hunk 5]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+    mpj *= 10**6
+
+    return mpj * gain
+
+
+def miles_per_gallon(env, veh_id=None, gain=.001):
+    """Calculate mpg of either a particular vehicle or the total average of all the vehilces.
+    Assumes vehicle is an average sized vehicle.
+    The power calculated here is the lower bound of the actual power consumed
+    by a vehicle.
```
Comment (Yasharzf) [inline flow/core/rewards.py:398]: params and returns?

File: flow/core/rewards.py [hunk 6]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+def miles_per_megajoule(env, veh_id=None, gain=.001):
+    """Calculate miles per mega-joule of either a particular vehicle or the total average of all the vehilces.
+    Assumes vehicle is an average sized vehicle.
+    The power calculated here is the lower bound of the actual power consumed
+    by a vehicle.
+    """
+    mpj = 0
+    counter = 0
+    if not isinstance(veh_id, list):
+        speed = env.k.vehicle.get_speed(veh_id)
```
Comment (Yasharzf) [inline flow/core/rewards.py:368]: what happens if veh_id is None?

File: flow/core/rewards.py [hunk 7]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+def miles_per_gallon(env, veh_id=None, gain=.001):
+    """Calculate mpg of either a particular vehicle or the total average of all the vehilces.
+    Assumes vehicle is an average sized vehicle.
+    The power calculated here is the lower bound of the actual power consumed
+    by a vehicle.
+    """
+    mpg = 0
+    counter = 0
+    if not isinstance(veh_id, list):
+        speed = env.k.vehicle.get_speed(veh_id)
```
Comment (Yasharzf) [inline flow/core/rewards.py:403]: same, if veh_id is None?

File: flow/core/rewards.py [hunk 8]
```diff
@@ -330,3 +330,93 @@ def energy_consumption(env, gain=.001):
...
+        gallons_per_s = env.k.vehicle.get_fuel_consumption(veh_id)
+        if gallons_per_s > 0 and speed >= 0.0:
+            # meters / gallon is (v * \delta t) / (gallons_s * \delta t)
+            mpg = speed / gallons_per_s
+    else:
+        for veh_id in env.k.vehicle.get_ids():
```
Comment (Yasharzf) [inline flow/core/rewards.py:409]: same

File: flow/core/rewards.py (hunk 2 above)
Comment (eugenevinitsky) [reply flow/core/rewards.py:354]: No, they're from Joy and are custom values

File: flow/core/rewards.py (hunk 6 above)
Comment (eugenevinitsky) [reply flow/core/rewards.py:368]: that'd be a higher up error. I can add an assert

**Pull request merged**